File: csharp/processor/Output/Serializer.cs
// @docspec:module {
//   id: "docspec-csharp-serializer",
//   name: "DocSpec Serializer",
//   description: "Serializes the DocSpec output model to JSON and writes docspec.json to the output directory.",
//   since: "3.0.0"
// }

using System.Text.Json;

namespace DocSpec.Analyzer.Output;

/// <summary>Write docspec.json output.</summary>
// [DocIntentional("Serialize the DocSpec output model to a pretty-printed JSON file")]
public class Serializer
{
    // [DocMethod(Since = "3.0.0")]
    // [DocIntentional("Serialize DocSpec output to pretty-printed JSON and write to docspec.json in the output directory")]
    public void Write(object data, string outputDir)
    {
        Directory.CreateDirectory(outputDir);
        var path = Path.Combine(outputDir, "docspec.json");
        var json = JsonSerializer.Serialize(data, new JsonSerializerOptions
        {
            WriteIndented = true,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
        });
        File.WriteAllText(path, json);
    }
}

File: go/processor/output/serializer.go
// Package output handles DocSpec JSON serialization.
//
// @docspec:module {
//   id: "docspec-go-output",
//   name: "Output Serializer",
//   description: "Serializes the assembled DocSpec model to a pretty-printed JSON file, the final step of the processor pipeline.",
//   since: "3.0.0"
// }
package output

import (
	"encoding/json"
	"os"
	"path/filepath"
)

// WriteJSON serializes data to a JSON file.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Writes the final docspec.json output with pretty-printed indentation"
func WriteJSON(data interface{}, outputDir, filename string) error {
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return err
	}

	path := filepath.Join(outputDir, filename)
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	return encoder.Encode(data)
}
